### ISSUE TITLE
Allow specifying a relationship-based JOIN on delete() structures

### ISSUE THREAD

**Issue by tgross35** [sqlalchemy/sqlalchemy#8130]
### Describe the use case

It doesn't currently seem possible to add a `.join()`  to a `delete()` object for multi-table deletes. Currently, the following is possible (implicit inner join):

```python
q=delete(TableA).filter(
            TableA.something == 'something',
            TableA.b_id == TableB.id,
            TableB.something == 'abcd',
        ).execution_options(synchronize_session="fetch")
q.stringify_dialect = 'mysql'
str(q)
```

```sql
DELETE FROM table_a USING table_a, table_b
WHERE 'something' = table_a.something
AND table_a.table_b_id = table_b.id
AND table_b.something = 'abcd'
```

But it would be convenient to be able to specify the join condition using an ORM relationship, and use the other types of joins possible (left).

### Databases / Backends / Drivers targeted

This at least works in mysql and mariadb, probably some others (unsure which)

Good examples are here: https://www.mysqltutorial.org/mysql-delete-join/

### Example Use

Rough example (untested)

```python
q=delete(TableA).join(TableA.b_relationship)filter(
            TableA.something == 'something',
            TableB.something == 'abcd',
        ).execution_options(synchronize_session="fetch")
```

```sql
DELETE TableA 
FROM TableA
JOIN TableB ON TableA .b_id = TableB.id
WHERE
    TableA.something = 'something'
AND
    TableB.something = 'abcd';
```

### Additional context

It is also possible to join on UPDATE - I am unsure what the current state of support for that is

https://www.mysqltutorial.org/mysql-update-join/

**Comment by zzzeek** [comment]
we support MySQLs join() for UPDATE, for DELETE it does not seem we have specific tests.

this would be a Core use case, not ORM.   the way it works is to pass join() directly to delete().  the relationship part you can get using orm.join().

```py
from sqlalchemy.orm import join

stmt = delete(join(ModelA, ModelB, ModelA.bs)) ...
```

right now it comes out like 

```
DELETE FROM a INNER JOIN b ON a.id = b.a_id USING a INNER JOIN b ON a.id = b.a_id, b
```

also it would be nice if both the update() and the delete() cases had some docs at https://docs.sqlalchemy.org/en/14/dialects/mysql.html#mysql-mariadb-sql-extensions

**Comment by zzzeek** [comment]
can that JOIN be in the USING clause  ?

**Comment by CaselIT** [comment]
Pg supports this also, but only on the using https://www.postgresql.org/docs/current/sql-delete.html

**Comment by tgross35** [comment]
Looks like the join condition can be after USING, this works for me

```sql
DELETE FROM T1 USING T1 LEFT JOIN T2
ON T1.id = T2.t1_id
WHERE T2.value = 'deleteme';
```

Per the docs, either of these are allowable and I think equivalent. The join just goes in table_references
```sql
DELETE [LOW_PRIORITY] [QUICK] [IGNORE]
    tbl_name[.*] [, tbl_name[.*]] ...
    FROM table_references
    [WHERE where_condition]
```

```sql
DELETE [LOW_PRIORITY] [QUICK] [IGNORE]
    FROM tbl_name[.*] [, tbl_name[.*]] ...
    USING table_references
    [WHERE where_condition]
```

**Comment by CaselIT** [comment]
> Pg supports this also, but only on the using https://www.postgresql.org/docs/current/sql-delete.html

But on pg side it seems that it can be used only to filter the rows. It does not lalow deleting from multiple tables, so probably not the same functionality

**Comment by tgross35** [comment]
> ```python
> from sqlalchemy.orm import join
> 
> stmt = delete(join(ModelA, ModelB, ModelA.bs)) ...
> ```

Does this distinguish between deleting from only ModelA and deleting from both ModelA and ModelB? Or am I missing context

**Comment by zzzeek** [comment]
well for DELETE it's whatever the database does I guess?  we just print a string

**Comment by tgross35** [comment]
Only from ModelA as-is (though the rendered sql syntax you sent is broken of course), I guess I was moreso wondering if there would be a way to specify a delete from both tables. Assuming that the python syntax is what the goal is, wasn't sure if the goal is to make that work or if that's just an existing workaround until better syntax could be added.

**Comment by iTrooz** [comment]
```
stmt = delete(TableA).join(TableB).filter(
    TableA.something == 'something',
    TableB.something == 'abcd'
)
```

Hey, is the following syntax being worked on ? If not, I would like to pick it up, would you accept a PR ?

**Comment by zzzeek** [comment]
yes we accept PRs, in this case it has to be pretty complete with good compile and round trip tests etc.  see test/sql/test_delete.py for where to go with this.

**Comment by sqla-tester** [comment]
**cjc0013** has proposed a fix for this issue in the **main** branch:

**Add explicit USING support to DELETE** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6731

**Comment by sqla-tester** [comment]
**cjc0013** has proposed a fix for this issue in the **main** branch:

**Add explicit USING support to DELETE** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6731
